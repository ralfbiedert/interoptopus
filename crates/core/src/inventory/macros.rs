/// Register an item.
#[macro_export]
macro_rules! extra_type {
    ($x:ty) => {{
        use $crate::lang::types::TypeInfo;

        |x: &mut $crate::inventory::Inventory| {
            <$x as TypeInfo>::register(x);
        }
    }};
}

/// Register an item.
#[macro_export]
macro_rules! function {
    ($x:ty) => {{
        |x: &mut $crate::inventory::Inventory| {
            <$x as $crate::lang::function::FunctionInfo>::register(x);
        }
    }};
}

/// Register an item.
#[macro_export]
macro_rules! service {
    ($x:ty) => {{
        |x: &mut $crate::inventory::Inventory| {
            <$x as $crate::lang::service::ServiceInfo>::register(x);
        }
    }};
}
