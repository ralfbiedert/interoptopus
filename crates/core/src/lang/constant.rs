use crate::lang::{Meta, Primitive, PrimitiveValue, Type};
use std::hash::{Hash, Hasher};

/// The value of a constant.
#[derive(Clone, Debug, PartialOrd, PartialEq)]
pub enum ConstantValue {
    Primitive(PrimitiveValue),
}

impl ConstantValue {
    pub(crate) fn fucking_hash_it_already<H: Hasher>(&self, h: &mut H) {
        match self {
            Self::Primitive(x) => match x {
                PrimitiveValue::Bool(x) => x.hash(h),
                PrimitiveValue::U8(x) => x.hash(h),
                PrimitiveValue::U16(x) => x.hash(h),
                PrimitiveValue::U32(x) => x.hash(h),
                PrimitiveValue::U64(x) => x.hash(h),
                PrimitiveValue::Usize(x) => x.hash(h),
                PrimitiveValue::I8(x) => x.hash(h),
                PrimitiveValue::I16(x) => x.hash(h),
                PrimitiveValue::I32(x) => x.hash(h),
                PrimitiveValue::I64(x) => x.hash(h),
                PrimitiveValue::Isize(x) => x.hash(h),
                PrimitiveValue::F32(x) => x.to_le_bytes().hash(h),
                PrimitiveValue::F64(x) => x.to_le_bytes().hash(h),
            },
        }
    }
}

/// A Rust `const` definition with a name and value, might become a `#define`.
#[derive(Clone, Debug, PartialOrd, PartialEq)]
pub struct Constant {
    name: String,
    value: ConstantValue,
    meta: Meta,
}

impl Constant {
    #[must_use]
    pub const fn new(name: String, value: ConstantValue, meta: Meta) -> Self {
        Self { name, value, meta }
    }

    #[must_use]
    pub fn name(&self) -> &str {
        &self.name
    }

    #[must_use]
    pub const fn value(&self) -> &ConstantValue {
        &self.value
    }

    #[must_use]
    pub const fn meta(&self) -> &Meta {
        &self.meta
    }

    /// Returns the type of this constant.
    #[must_use]
    pub const fn the_type(&self) -> Type {
        match &self.value {
            ConstantValue::Primitive(x) => Type::Primitive(match x {
                PrimitiveValue::Bool(_) => Primitive::Bool,
                PrimitiveValue::U8(_) => Primitive::U8,
                PrimitiveValue::U16(_) => Primitive::U16,
                PrimitiveValue::U32(_) => Primitive::U32,
                PrimitiveValue::U64(_) => Primitive::U64,
                PrimitiveValue::Usize(_) => Primitive::Usize,
                PrimitiveValue::I8(_) => Primitive::I8,
                PrimitiveValue::I16(_) => Primitive::I16,
                PrimitiveValue::I32(_) => Primitive::I32,
                PrimitiveValue::I64(_) => Primitive::I64,
                PrimitiveValue::Isize(_) => Primitive::Isize,
                PrimitiveValue::F32(_) => Primitive::F32,
                PrimitiveValue::F64(_) => Primitive::F64,
            }),
        }
    }
}
