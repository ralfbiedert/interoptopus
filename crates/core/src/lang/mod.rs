//! Metadata for the Rust and C representation of language items.
//!
//! Normal users of Interoptopus probably won't have to concern
//! themselves with any of the items in this module.

use crate::backend::{capitalize_first_letter, types_from_type_recursive};
use crate::pattern::TypePattern;
use std::collections::HashSet;

use crate::pattern::callback::AsyncCallback;
use crate::pattern::result::ResultType;
pub use array::Array;
pub use composite::{Composite, Field, Layout, Opaque, Representation};
pub use constant::{Constant, ConstantValue};
pub use enums::{Enum, Variant, VariantKind};
pub use fnpointer::FnPointer;
pub use function::{Function, Parameter, Signature, SugaredReturnType};
pub use info::{ConstantInfo, FunctionInfo, TypeInfo};
pub use meta::{Docs, Meta, Visibility};
pub use primitive::{Primitive, PrimitiveValue};

mod array;
mod composite;
mod constant;
mod enums;
mod fnpointer;
mod function;
mod info;
mod meta;
mod primitive;

/// A type that can exist at the FFI boundary.
#[derive(Clone, Debug, Ord, PartialOrd, Eq, PartialEq, Hash)]
pub enum Type {
    Primitive(Primitive),
    Array(Array),
    Enum(Enum),
    Opaque(Opaque),
    Composite(Composite),
    FnPointer(FnPointer),
    ReadPointer(Box<Type>),
    ReadWritePointer(Box<Type>),
    /// Special patterns with primitives existing on C-level but special semantics.
    /// useful to higher level languages.
    Pattern(TypePattern),
}

impl Default for Type {
    fn default() -> Self {
        Self::Primitive(Primitive::Void)
    }
}

impl Type {
    #[must_use]
    pub const fn size_of(&self) -> usize {
        match self {
            Self::Primitive(p) => match p {
                Primitive::Void => 0,
                Primitive::Bool => 1,
                Primitive::U8 => 1,
                Primitive::U16 => 2,
                Primitive::U32 => 4,
                Primitive::U64 => 8,
                Primitive::Usize => std::mem::size_of::<usize>(),
                Primitive::I8 => 1,
                Primitive::I16 => 2,
                Primitive::I32 => 4,
                Primitive::I64 => 8,
                Primitive::Isize => std::mem::size_of::<isize>(),
                Primitive::F32 => 4,
                Primitive::F64 => 8,
            },
            // TODO
            _ => 999,
        }
    }

    #[must_use]
    pub fn align_of(&self) -> usize {
        unimplemented!()
    }

    #[must_use]
    pub const fn void() -> Self {
        Self::Primitive(Primitive::Void)
    }

    #[must_use]
    pub fn to_type(&self) -> Self {
        self.clone()
    }

    /// Produces a name unique for that type with respect to this library.
    ///
    /// The name here is supposed to uniquely determine a type relative to a library [`Inventory`](crate::inventory::Inventory).
    ///
    /// Backends may instead match on the [`Type`] variant and determine a more appropriate
    /// name on a case-by-case basis; including changing a name entirely.
    #[must_use]
    pub fn name_within_lib(&self) -> String {
        match self {
            Self::Primitive(x) => x.rust_name().to_string(),
            Self::Enum(x) => x.rust_name().to_string(),
            Self::Opaque(x) => x.rust_name().to_string(),
            Self::Composite(x) => x.rust_name().to_string(),
            Self::FnPointer(x) => x.rust_name(),
            Self::ReadPointer(x) => format!("ConstPtr{}", capitalize_first_letter(x.name_within_lib().as_str())),
            Self::ReadWritePointer(x) => format!("MutPtr{}", capitalize_first_letter(x.name_within_lib().as_str())),
            Self::Pattern(x) => match x {
                TypePattern::Bool => "Bool".to_string(),
                _ => x.fallback_type().name_within_lib(),
            },
            Self::Array(x) => x.rust_name(),
        }
    }

    /// Lists all _other_ types this type refers to.
    #[must_use]
    pub fn embedded_types(&self) -> Vec<Self> {
        let mut hash_set: HashSet<Self> = HashSet::new();

        types_from_type_recursive(self, &mut hash_set);

        hash_set.remove(self);
        hash_set.iter().cloned().collect()
    }

    /// If this were a pointer, tries to deref it and return the inner type.
    #[must_use]
    pub fn try_deref_pointer(&self) -> Option<&Self> {
        match self {
            Self::Primitive(_) => None,
            Self::Enum(_) => None,
            Self::Opaque(_) => None,
            Self::Composite(_) => None,
            Self::FnPointer(_) => None,
            Self::ReadPointer(x) => Some(x.as_ref()),
            Self::ReadWritePointer(x) => Some(x.as_ref()),
            Self::Pattern(_) => None,
            Self::Array(_) => None,
        }
    }

    /// Convenience method attempting to convert the contained type as a composite.
    #[must_use]
    pub const fn as_composite_type(&self) -> Option<&Composite> {
        match self {
            Self::Composite(x) => Some(x),
            _ => None,
        }
    }

    /// Convenience method attempting to convert the contained type as an opaque.
    #[must_use]
    pub const fn as_opaque_type(&self) -> Option<&Opaque> {
        match self {
            Self::Opaque(x) => Some(x),
            _ => None,
        }
    }

    #[must_use]
    pub const fn as_result_type(&self) -> Option<&ResultType> {
        match self {
            Self::Pattern(TypePattern::Result(x)) => Some(x),
            _ => None,
        }
    }

    /// Convenience method attempting to convert the contained type as an Async callback.
    #[must_use]
    pub const fn as_async_callback(&self) -> Option<&AsyncCallback> {
        match self {
            Self::Pattern(TypePattern::AsyncCallback(x)) => Some(x),
            _ => None,
        }
    }

    /// Convenience method attempting to get the pointer target of a contained type.
    #[must_use]
    pub const fn pointer_target(&self) -> Option<&Self> {
        match self {
            Self::ReadPointer(x) => Some(x),
            Self::ReadWritePointer(x) => Some(x),
            _ => None,
        }
    }

    /// Convenience method attempting to convert the contained type as a composite.
    #[must_use]
    pub const fn as_result(&self) -> Option<&ResultType> {
        match self {
            Self::Pattern(TypePattern::Result(x)) => Some(x),
            _ => None,
        }
    }

    /// Checks if this is a [`Primitive::Void`].
    #[must_use]
    pub const fn is_void(&self) -> bool {
        matches!(self, Self::Primitive(Primitive::Void))
    }

    /// Returns the namespace of the type.
    #[must_use]
    pub fn namespace(&self) -> Option<&str> {
        match self {
            Self::Array(t) => t.the_type().namespace(),
            Self::Enum(t) => Some(t.meta().module()),
            Self::Opaque(t) => Some(t.meta().module()),
            Self::Composite(t) => Some(t.meta().module()),
            Self::Pattern(TypePattern::NamedCallback(t)) => Some(t.meta().module()),
            _ => None,
        }
    }
}
