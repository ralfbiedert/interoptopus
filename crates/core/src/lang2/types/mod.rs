mod array;
mod enums;
mod fnptr;
mod pattern;
mod primitive;
mod std;
mod structs;
mod wire;

use crate::lang2::function::Signature;
use crate::lang2::meta::{Docs, Emission, Visibility};

use crate::inventory2::TypeId;
pub use array::Array;
pub use enums::{Enum, Variant};
pub use pattern::TypePattern;
pub use primitive::{Primitive, PrimitiveValue};
pub use structs::{Field, Struct};
pub use wire::WireOnly;

pub trait TypeInfo {
    fn id() -> TypeId;
}

pub enum TypeKind {
    Array(Array),
    Primitive(Primitive),
    Struct(Struct),
    Enum(Enum),
    FnPointer(Signature),
    ReadPointer(TypeId),
    ReadWritePointer(TypeId),
    /// A type that may only be observed behind a pointer.
    Opaque,
    /// A type that can only appear inside a `Wire<T>`
    WireOnly(WireOnly),
    TypePattern(TypePattern),
}

pub struct Type {
    pub name: String,
    pub visibility: Visibility,
    pub docs: Docs,
    pub emission: Emission,
    pub kind: TypeKind,
}

/// How a struct or enum is laid out in memory.
#[derive(Clone, Copy, Debug, Ord, PartialOrd, Eq, PartialEq, Hash)]
pub enum Layout {
    C,
    Transparent,
    Packed,
    Opaque,
    /// For use with enum discriminant.
    Primitive(crate::lang::Primitive),
}

/// How a type is represented in memory.
#[derive(Clone, Copy, Debug, Ord, PartialOrd, Eq, PartialEq, Hash)]
pub struct Repr {
    layout: crate::lang::Layout,
    alignment: Option<usize>,
}
