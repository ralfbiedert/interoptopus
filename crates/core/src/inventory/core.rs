use crate::inventory::forbidden::FORBIDDEN_NAMES;
use crate::lang::util::{extract_namespaces_from_types, extract_wire_types_from_functions, holds_opaque_without_ref, types_from_functions_types};
use crate::lang::{Constant, Function, Type};
use crate::pattern::LibraryPattern;
use std::collections::HashSet;

/// Tells the [`InventoryBuilder`] what to register.
///
/// Most users won't need to touch this enum directly, as its variants are usually created via the [`function`](crate::function), [`constant`](crate::constant), [`extra_type`](crate::extra_type) and [`pattern`](crate::pattern!) macros.
#[derive(Debug)]
pub enum Symbol {
    Function(Function),
    Constant(Constant),
    Type(Type),
    Pattern(LibraryPattern),
}

/// Produces an [`Inventory`] inside your inventory function, **start here**.🔥
///
/// # Example
///
/// Define an inventory function containing a function, constant, and an extra type.
/// This function can be called from your unit tests and the returned [`Inventory`] used to create bindings.
///
/// ```rust
/// use interoptopus::{function, constant, extra_type, pattern, ffi_function, ffi_constant, ffi_type};
/// use interoptopus::inventory::{Inventory, InventoryBuilder};
///
/// // First, define some items our DLL uses or needs.
///
/// #[ffi_function]
/// pub fn primitive_void() { }
///
/// #[ffi_constant]
/// pub const MY_CONSTANT: u32 = 123;
///
/// #[ffi_type]
/// pub struct ExtraType<T> {
///     x: T
/// }
///
/// // Then list all items for which to generate bindings. Call this function
/// // from another crate or unit test and feed the `Library` into a backend to
/// // generate bindings for a specific language.
/// pub fn my_inventory() -> Inventory {
///     Inventory::builder()
///         .register(function!(primitive_void))
///         .register(constant!(MY_CONSTANT))
///         .register(extra_type!(ExtraType<f32>))
///         .validate()
///         .build()
/// }
/// ```
#[derive(Default, Debug)]
pub struct InventoryBuilder {
    functions: Vec<Function>,
    extra_types: Vec<Type>,
    constants: Vec<Constant>,
    patterns: Vec<LibraryPattern>,
    allow_reserved_names: bool,
}

impl InventoryBuilder {
    /// Start creating a new library.
    #[must_use]
    const fn new() -> Self {
        Self { functions: Vec::new(), extra_types: Vec::new(), /*wire_types: Vec::new(),*/ constants: Vec::new(), patterns: Vec::new(), allow_reserved_names: false }
    }

    /// Registers a symbol.
    ///
    /// Call this with the result of a [`function`](crate::function), [`constant`](crate::constant), [`extra_type`](crate::extra_type) or [`pattern`](crate::pattern!) macro,
    /// see the example above.
    #[must_use]
    pub fn register(mut self, s: Symbol) -> Self {
        match s {
            Symbol::Function(x) => self.functions.push(x),
            Symbol::Constant(x) => self.constants.push(x),
            Symbol::Type(x) => self.extra_types.push(x),
            Symbol::Pattern(x) => {
                match &x {
                    LibraryPattern::Service(x) => {
                        self.functions.push(x.destructor().clone());
                        self.functions.extend(x.constructors().iter().cloned());
                        self.functions.extend(x.methods().iter().cloned());
                    }
                    LibraryPattern::Builtins(x) => {
                        self.functions.extend(x.functions().iter().cloned());
                    }
                }
                self.patterns.push(x);
            }
        }

        self
    }

    /// Does additional sanity checking, highly recommended.
    ///
    /// This method tries to detect FFI issues that are hard to detect otherwise, and would
    /// cause issues in any backend.
    ///
    /// # Panics
    ///
    /// If a function, type, or pattern is detected that doesn't make sense in interop
    /// generation, a panic will be raised.
    #[must_use]
    pub fn validate(self) -> Self {
        // Check for opaque parameters and return values
        for x in &self.functions {
            let has_opaque_param = x.signature().params().iter().any(|x| holds_opaque_without_ref(x.the_type()));
            assert!(!has_opaque_param, "Function `{}` has a (nested) opaque parameter. This can cause UB.", x.name());

            let has_opaque_rval = holds_opaque_without_ref(x.signature().rval());
            assert!(!has_opaque_rval, "Function `{}` has a (nested) opaque return value. This can cause UB.", x.name());
        }

        if !self.allow_reserved_names {
            validate_symbol_names(&self.functions, &self.extra_types);
        }

        self
    }

    /// Allows reserved names for inventory items.
    ///
    /// When set, you may use reserved names for your functions
    /// and fields, for example, having a field called `public`.
    ///
    /// When not set, [`self.validate`] may panic if it detects
    /// such items.
    #[must_use]
    pub fn allow_reserved_names(mut self) -> Self {
        self.allow_reserved_names = true;
        self
    }

    /// Produce the [`Inventory`].
    #[must_use]
    pub fn build(self) -> Inventory {
        Inventory::new(self.functions, self.constants, self.patterns, self.extra_types.as_slice())
    }
}

/// Holds FFI-relevant items, produced via [`InventoryBuilder`], ingested by backends.
#[derive(Clone, Debug, PartialOrd, PartialEq, Default)]
pub struct Inventory {
    functions: Vec<Function>,
    /// FFI types and wire payload types.
    c_types: Vec<Type>,
    /// These are the types explicitly marked as `Wire<T>` in function declarations, we extract them here so we can
    /// rebuild the chain of custody and generate all appropriate types.
    /// Other wired types contained within these listed ones are already added to `c_types` as `WirePayload` types.
    wire_types: Vec<Type>,
    constants: Vec<Constant>,
    patterns: Vec<LibraryPattern>,
    namespaces: Vec<String>,
}

/// References to items contained within an [`Inventory`].
///
/// This enum is primarily used by functions such as [`Inventory::filter`] that need to handle
/// items of various types.
#[derive(Clone, Debug, PartialEq)]
pub enum InventoryItem<'a> {
    Function(&'a Function),
    CType(&'a Type),
    WireType(&'a Type),
    Constant(&'a Constant),
    Pattern(&'a LibraryPattern),
    Namespace(&'a str),
}

fn dedup<T: std::hash::Hash + Eq>(v: Vec<T>) -> Vec<T> {
    v.into_iter().collect::<HashSet<T>>().into_iter().collect::<Vec<T>>()
}

// fn dedup_sort<T: Ord>(mut v: Vec<T>) -> Vec<T> {
//     v.sort();
//     v.dedup();
//     v
// }

impl Inventory {
    /// Produce a new inventory for the given functions, constants and patterns.
    ///
    /// Type information will be automatically derived from the used fields and parameters.
    pub(crate) fn new(functions: Vec<Function>, constants: Vec<Constant>, patterns: Vec<LibraryPattern>, extra_types: &[Type]) -> Self {
        let mut c_types = types_from_functions_types(&functions, extra_types);
        let mut namespaces = HashSet::new();

        // Extract namespace information
        extract_namespaces_from_types(&c_types, &mut namespaces);
        namespaces.extend(functions.iter().map(|x| x.meta().module().to_string()));
        namespaces.extend(constants.iter().map(|x| x.meta().module().to_string()));

        let mut namespaces = namespaces.iter().cloned().collect::<Vec<String>>();
        namespaces.sort();

        let wire_types = dedup(extract_wire_types_from_functions(&functions));

        // Dont sort functions
        // functions.sort();

        c_types.sort();
        // constants.sort(); TODO: do sort constants (issue with Ord and float values ...)

        Self { functions, c_types, wire_types, constants, patterns, namespaces }
    }

    /// Helper to debug the inventory.
    pub fn debug(&self) {
        eprintln!("✅ Inventory check");
        for ns in &self.namespaces {
            eprintln!("💳 {ns}");
        }
        for cs in &self.constants {
            eprintln!("👮🏼‍♀️ {}", cs.name());
        }
        for f in &self.functions {
            eprintln!(
                "🧵 {}.{}({}) -> {}",
                f.meta().module(),
                f.name(),
                f.signature().params().iter().map(|p| p.the_type().name_within_lib()).collect::<Vec::<_>>().join(","),
                f.signature().rval().name_within_lib()
            );
        }
        for c in &self.c_types {
            eprintln!("🎉 {}", c.name_within_lib());
        }
        for w in &self.wire_types {
            eprintln!("🔧 {} '{}'", w.name_within_lib(), w.namespace().unwrap_or_default());
        }
        // for p in &self.patterns {
        //     eprintln!("🧵 {}", p.fallback_type().name_within_lib());
        // }
    }

    /// Returns a new [`InventoryBuilder`], start here.
    #[must_use]
    pub const fn builder() -> InventoryBuilder {
        InventoryBuilder::new()
    }

    /// Return all functions registered.
    #[must_use]
    pub fn functions(&self) -> &[Function] {
        &self.functions
    }

    /// Returns all found types; this includes types directly used in fields and parameters, and
    /// all their recursive constituents.
    #[must_use]
    pub fn c_types(&self) -> &[Type] {
        &self.c_types
    }

    /// Returns initial wire types; this includes wire types directly used in function parameters.
    /// To find all transitively wired types, look for `wire_domain_types()`
    #[must_use]
    pub fn wire_types(&self) -> &[Type] {
        &self.wire_types
    }

    /// Returns wire payload types; this includes types `T` directly and indirectly used in `Wire<T>` function parameters.
    #[must_use]
    #[allow(clippy::redundant_closure_for_method_calls)]
    pub fn wire_domain_types(&self) -> Vec<Type> {
        let trans_types = self
            .wire_types
            .iter()
            .flat_map(|wt| match wt {
                Type::Wire(w) => w
                    .fields()
                    .iter()
                    .filter(|&f| matches!(f.the_type(), Type::Composite(_)))
                    .map(|f| f.the_type())
                    .filter(|ty| self.c_types.contains(ty)),
                _ => panic!("What's a non-wired type doing here?"),
            })
            .cloned()
            .collect::<HashSet<Type>>();
        trans_types.into_iter().collect()
    }

    /// Return all registered constants.
    #[must_use]
    pub fn constants(&self) -> &[Constant] {
        &self.constants
    }

    /// Return all known namespaces.
    #[must_use]
    pub fn namespaces(&self) -> &[String] {
        &self.namespaces
    }

    /// Return all registered [`LibraryPattern`]. In contrast, [`TypePattern`](crate::pattern::TypePattern)
    /// will be found inside the types returned via [`c_types()`](Self::c_types).
    #[must_use]
    pub fn patterns(&self) -> &[LibraryPattern] {
        &self.patterns
    }

    /// Return a new [`Inventory`] filtering items by a predicate.
    ///
    /// Useful for removing duplicate symbols when generating bindings split across multiple files.
    ///
    /// # Examples
    ///
    /// Here we filter an inventory, keeping only types, removing all other items.
    ///
    /// ```rust
    /// # use interoptopus::inventory::{Inventory, InventoryItem};
    /// #
    /// # let inventory = Inventory::default();
    /// #
    /// let filtered = inventory.filter(|x| {
    ///     match x {
    ///         InventoryItem::CType(_) => true,
    ///         _ => false,
    ///     }
    /// });
    /// ```
    #[must_use]
    pub fn filter<P: FnMut(InventoryItem) -> bool>(&self, mut predicate: P) -> Self {
        let functions: Vec<Function> = self.functions.iter().filter(|x| predicate(InventoryItem::Function(x))).cloned().collect();
        let c_types: Vec<Type> = self.c_types.iter().filter(|x| predicate(InventoryItem::CType(x))).cloned().collect();
        let wire_types: Vec<Type> = self.wire_types.iter().filter(|x| predicate(InventoryItem::WireType(x))).cloned().collect();
        let constants: Vec<Constant> = self.constants.iter().filter(|x| predicate(InventoryItem::Constant(x))).cloned().collect();
        let patterns: Vec<LibraryPattern> = self.patterns.iter().filter(|x| predicate(InventoryItem::Pattern(x))).cloned().collect();
        let namespaces: Vec<String> = self.namespaces.iter().filter(|x| predicate(InventoryItem::Namespace(x))).cloned().collect();

        Self { functions, c_types, wire_types, constants, patterns, namespaces }
    }
}

fn validate_symbol_names(functions: &[Function], c_types: &[Type]) {
    // Check function names and parameter names.
    for func in functions {
        let name = func.name().to_lowercase();
        assert!(!FORBIDDEN_NAMES.contains(&name.as_str()), "Function `{name}` has a forbidden name that might cause issues in other languages.");

        for param in func.signature().params() {
            let param_name = param.name().to_lowercase();
            assert!(
                !FORBIDDEN_NAMES.contains(&param_name.as_str()),
                "Parameter `{param_name}` in function `{name}` has a forbidden name that might cause issues in other languages."
            );
        }
    }

    // Check type names and field/variant names.
    for ctype in c_types {
        match ctype {
            Type::Composite(composite) => {
                let type_name = composite.rust_name();
                assert!(!FORBIDDEN_NAMES.contains(&type_name), "Type `{type_name}` has a forbidden name that might cause issues in other languages.");
                for field in composite.fields() {
                    let field_name = field.name();
                    assert!(
                        !FORBIDDEN_NAMES.contains(&field_name),
                        "Field `{field_name}` in type `{type_name}` has a forbidden name that might cause issues in other languages."
                    );
                }
            }
            Type::Enum(enum_type) => {
                let type_name = enum_type.rust_name();
                // Inline the format parameter using Rust's string interpolation.
                assert!(!FORBIDDEN_NAMES.contains(&type_name), "Enum `{type_name}` has a forbidden name that might cause issues in other languages.");
                for variant in enum_type.variants() {
                    let variant_name = variant.name();
                    assert!(
                        !FORBIDDEN_NAMES.contains(&variant_name),
                        "Variant `{variant_name}` in enum `{type_name}` has a forbidden name that might cause issues in other languages."
                    );
                }
            }
            Type::Opaque(opaque) => {
                let type_name = opaque.rust_name();
                assert!(!FORBIDDEN_NAMES.contains(&type_name), "Opaque type `{type_name}` has a forbidden name that might cause issues in other languages.");
            }
            _ => {}
        }
    }
}
