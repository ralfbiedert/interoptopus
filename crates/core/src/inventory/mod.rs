//! Helpers for dealing with an FFI inventory.
//!

use crate::pattern::LibraryPattern;

pub(crate) mod core;
pub(crate) mod forbidden;

use crate::lang::Function;
pub use core::{Inventory, InventoryBuilder, InventoryItem, Symbol};

/// Returns all functions not belonging to a [`service`](crate::pattern::service) pattern.
///
/// Useful in backends like Python that can fully encapsulate services and should not expose their
/// raw methods in the main namespace.
#[must_use]
pub fn non_service_functions(inventory: &Inventory) -> Vec<&Function> {
    let mut service_methods = vec![];
    for pattern in inventory.patterns() {
        match pattern {
            LibraryPattern::Service(service) => {
                service_methods.extend_from_slice(service.methods());
                service_methods.extend_from_slice(service.constructors());
                service_methods.push(service.destructor().clone());
            }
            LibraryPattern::Builtins(_) => {}
        }
    }

    inventory.functions().iter().filter(|&x| !service_methods.contains(x)).collect()
}

/// Create a single [`Inventory`] from a number of individual inventories.
///
/// This function can be useful when your FFI crate exports different sets of
/// symbols (e.g., _core_ and _extension_ functions) and you want to create different
/// bindings based on some compile target or configuration
///
/// # Example
///
/// ```
/// # mod my_crate {
/// #     use interoptopus::inventory::Inventory;
/// #     pub fn inventory_core() -> Inventory { Inventory::default() }
/// #     pub fn inventory_ext() -> Inventory { Inventory::default() }
/// # }
///
/// use interoptopus::inventory::merge_inventories;
///
/// let inventories = [
///     my_crate::inventory_core(),
///     my_crate::inventory_ext()
/// ];
///
/// merge_inventories(&inventories);
/// ```
#[must_use]
pub fn merge_inventories(inventories: &[Inventory]) -> Inventory {
    let mut functions = Vec::new();
    let mut constants = Vec::new();
    let mut patterns = Vec::new();
    let mut types = Vec::new();

    for inventory in inventories {
        functions.extend_from_slice(inventory.functions());
        constants.extend_from_slice(inventory.constants());
        patterns.extend_from_slice(inventory.patterns());
        types.extend_from_slice(inventory.c_types());
    }

    Inventory::new(functions, constants, patterns, types.as_slice())
}
