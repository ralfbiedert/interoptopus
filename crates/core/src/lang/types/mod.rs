mod array;
mod enums;
mod fnptr;
mod pattern;
mod primitive;
mod std;
mod structs;
mod wire;

use crate::lang::function::Signature;
use crate::lang::meta::{Docs, Emission, Visibility};
use ::std::io::{Read, Write};

use crate::inventory::{Inventory, TypeId};
pub use array::Array;
pub use enums::{Enum, Variant, VariantKind};
pub use pattern::TypePattern;
pub use primitive::{Primitive, PrimitiveValue};
pub use structs::{Field, Struct};
pub use wire::WireOnly;

pub trait TypeInfo {
    // Flags that govern where a type can be used. Allows
    // us to emit `const {}` guard blocks to produce compile time
    // errors.
    const WIRE_SAFE: bool;
    const RAW_SAFE: bool;
    const ASYNC_SAFE: bool;
    const SERVICE_SAFE: bool;
    const SERVICE_CTOR_SAFE: bool;

    // Basic helpers producing infos about the type
    fn id() -> TypeId;
    fn kind() -> TypeKind;
    fn ty() -> Type;

    // Registers the type and all dependents in an Inventory.
    fn register(inventory: &mut Inventory);

    // Utilities for (de)serializing an instance of this type. These must be
    // properly implemented if WIRE_SAFE is true. Otherwise these
    // should panic.
    fn write(&self, out: &mut impl Write) -> Result<(), SerializationError>;
    fn read(input: &mut impl Read) -> Result<Self, SerializationError>
    where
        Self: Sized;
    fn live_size(&self) -> usize;
}

#[derive(Clone, Debug, PartialEq, Eq, Hash)]
pub enum TypeKind {
    Array(Array),
    Primitive(Primitive),
    Struct(Struct),
    Enum(Enum),
    FnPointer(Signature),
    ReadPointer(TypeId),
    Service,
    ReadWritePointer(TypeId),
    /// A type that may only be observed behind a pointer.
    Opaque,
    /// A type that can only appear inside a `Wire<T>`
    WireOnly(WireOnly),
    TypePattern(TypePattern),
}

#[derive(Clone, Debug, PartialEq, Eq, Hash)]
pub struct Type {
    pub name: String,
    pub visibility: Visibility,
    pub docs: Docs,
    pub emission: Emission,
    pub kind: TypeKind,
}

/// How a struct or enum is laid out in memory.
#[derive(Clone, Copy, Debug, PartialOrd, Eq, PartialEq, Hash)]
pub enum Layout {
    C,
    Transparent,
    Packed,
    Opaque,
    /// For use with enum discriminant.
    Primitive(Primitive),
}

/// How a type is represented in memory.
#[derive(Clone, Copy, Debug, PartialOrd, Eq, PartialEq, Hash)]
pub struct Repr {
    pub layout: Layout,
    pub alignment: Option<usize>,
}

impl Repr {
    #[must_use]
    pub fn c() -> Self {
        Self { layout: Layout::C, alignment: None }
    }

    #[must_use]
    pub fn u32() -> Self {
        Self { layout: Layout::Primitive(Primitive::U32), alignment: None }
    }
}

#[track_caller]
pub const fn assert_wire_safe<T: TypeInfo>() {
    assert!(T::WIRE_SAFE);
}

#[track_caller]
pub const fn assert_raw_safe<T: TypeInfo>() {
    assert!(T::RAW_SAFE, "This type cannot be safely passed over FFI boundaries.");
}

#[track_caller]
pub const fn assert_async_safe<T: TypeInfo>() {
    assert!(T::ASYNC_SAFE);
}

#[track_caller]
pub const fn assert_service_safe<T: TypeInfo>() {
    assert!(T::SERVICE_SAFE);
}

#[track_caller]
pub const fn assert_service_ctor_safe<T: TypeInfo>() {
    assert!(T::SERVICE_CTOR_SAFE, "This method looks like a constructor, but does not return ffi::Result<Self, _>");
}

/// If a wire transfer goes wrong.
// @todo play with implementing it as a struct?
#[derive(Debug)]
pub enum SerializationError {
    Io(::std::io::Error),
    InvalidData(String),
    InvalidDiscriminant(String, usize),
}

impl From<::std::io::Error> for SerializationError {
    fn from(e: ::std::io::Error) -> Self {
        Self::Io(e)
    }
}
