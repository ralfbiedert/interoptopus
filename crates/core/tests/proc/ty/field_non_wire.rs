use interoptopus::ffi;
use interoptopus::inventory::{Inventory, TypeId};
use interoptopus::lang::types::{Type, TypeInfo, TypeKind};

// This simulates the error diagnostic if a type implemented `TypeInfo` manually
// but forgot to impl WireIO.

struct Bar;

impl TypeInfo for Bar {
    const WIRE_SAFE: bool = true;
    const RAW_SAFE: bool = false;
    const ASYNC_SAFE: bool = false;
    const SERVICE_SAFE: bool = false;
    const SERVICE_CTOR_SAFE: bool = false;

    fn id() -> TypeId {
        TypeId::new(0x0)
    }
    fn kind() -> TypeKind {
        todo!()
    }
    fn ty() -> Type {
        todo!()
    }
    fn register(_: &mut Inventory) {}
}

#[ffi]
struct Service {
    x: f32,
    y: Bar,
}

fn main() {}
