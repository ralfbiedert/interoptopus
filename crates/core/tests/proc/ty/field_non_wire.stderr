error[E0277]: the trait bound `Bar: WireIO` is not satisfied
  --> tests/proc/ty/field_non_wire.rs:32:5
   |
32 |     y: Bar,
   |     ^^^^^^ the trait `WireIO` is not implemented for `Bar`
   |
   = help: the following other types implement trait `WireIO`:
             &T
             &mut T
             ()
             *const T
             *mut T
             ApiVersion
             CStrPtr<'_>
             HashMap<K, V, S>
           and $N others
   = help: see issue #48214
help: add `#![feature(trivial_bounds)]` to the crate attributes to enable
   |
 1 + #![feature(trivial_bounds)]
   |
